src/backend/distributed/commands/index.c

```diff
@@ -164,29 +175,230 @@ PreprocessIndexStmt(Node *node, const char *createIndexCommand)
+			DDLJob *ddlJob = palloc0(sizeof(DDLJob));
+			ddlJob->targetRelationId = relationId;
+			ddlJob->concurrentIndexCmd = createIndexStatement->concurrent;
+			ddlJob->commandString = createIndexCommand;
+			ddlJob->taskList = CreateIndexTaskList(relationId, createIndexStatement);
 
-			/* if index does not exist, send the command to workers */
-			if (!OidIsValid(indexRelationId))
-			{
-				DDLJob *ddlJob = palloc0(sizeof(DDLJob));
-				ddlJob->targetRelationId = relationId;
-				ddlJob->concurrentIndexCmd = createIndexStatement->concurrent;
-				ddlJob->commandString = createIndexCommand;
-				ddlJob->taskList = CreateIndexTaskList(relationId, createIndexStatement);
+			ddlJobs = list_make1(ddlJob);
+		}
+	}
+
+	return ddlJobs;
+}
+
+
+/*
+ * GenerateDefaultIndexName generates default index name for the index to be
+ * created on relation as postgres would do.
+ */
+static char *
+GenerateDefaultIndexName(Relation relation, IndexStmt *createIndexStatement)
+{
+	List *indexParams = createIndexStatement->indexParams;
+	List *indexIncludingParams = createIndexStatement->indexIncludingParams;
+	List *allIndexParams = list_concat(list_copy(indexParams),
+									   list_copy(indexIncludingParams));
+
+	int numberOfAttributes = list_length(allIndexParams);
+	if (numberOfAttributes == 0 || numberOfAttributes > INDEX_MAX_KEYS)
+	{
+		/* cannot generate a default name for the index */
+		return NULL;
+	}
+
+	List *indexColNames = ChooseIndexColumnNames(allIndexParams);
+	Oid namespaceId = RelationGetNamespace(relation);
+	char *indexName = ChooseIndexName(RelationGetRelationName(relation),
+									  namespaceId,
+									  indexColNames,
+									  createIndexStatement->excludeOpNames,
+									  createIndexStatement->primary,
+									  createIndexStatement->isconstraint);
+
+	return indexName;
+}
 
-				ddlJobs = list_make1(ddlJob);
+
+/*
+ * Select the actual names to be used for the columns of an index, given the
+ * list of IndexElems for the columns.  This is mostly about ensuring the
+ * names are unique so we don't get a conflicting-attribute-names error.
+ *
+ * Returns a List of plain strings (char *, not String nodes).
+ *
+ * (Directly copied from postgres/commands/indexcmds.c)
+ */
+static List *
+ChooseIndexColumnNames(List *indexElems)
+{
+	List *result = NIL;
+	ListCell *lc;
+
+	foreach(lc, indexElems)
+	{
+		IndexElem *ielem = (IndexElem *) lfirst(lc);
+		const char *origname;
+		int i;
+		char buf[NAMEDATALEN];
+
+		/* Get the preliminary name from the IndexElem */
+		if (ielem->indexcolname)
+		{
+			origname = ielem->indexcolname; /* caller-specified name */
+		}
+		else if (ielem->name)
+		{
+			origname = ielem->name; /* simple column reference */
+		}
+		else
+		{
+			origname = "expr";  /* default name for expression */
+		}
+
+		/* If it conflicts with any previous column, tweak it */
+		const char *curname = origname;
+		for (i = 1;; i++)
+		{
+			ListCell *lc2;
+			char nbuf[32];
```

Comment:
I wonder is this hardcoded `32` related with default value of `INDEX_MAX_KEYS` ? (I guess no)